### PR TITLE
Feature/no more anuvu/squashfs dep

### PR DESCRIPTION
**Pull request by smoser**
stacker was only using anuvu/squashfs import to get the size of the squashfs file (from its superblock, rather than from 'stat').

That go library used libsquashfs-tools-ng as a c-go dependency, which meant a build and runtime dependency on it.

The change here is to drop that import and instead use some native go.

### CODE REVIEW

**Comment by smoser** [comment]
Given the superblock.go is copied from diskfs, I wonder if I need to add something like this:

```diff
diff --git a/pkg/squashfs/superblock.go b/pkg/squashfs/superblock.go
index a90ab5b..c16e1ee 100644
--- a/pkg/squashfs/superblock.go
+++ b/pkg/squashfs/superblock.go
@@ -1,3 +1,12 @@
+/*
+Copyright (c) 2017 Avi Deitcher
+
+This file was initially copied from go-diskfs [1].
+The copied portion is licensed under the terms of the MIT license [2].
+ [1] https://github.com/diskfs/go-diskfs/filesystem/squashfs/superblock.go
+ [2] https://opensource.org/licenses/MI
+*/
+
 package squashfs
 
 import (

```

**Comment by tych0** [comment]
My only thought is that it would be nice to import superblock.go instead of copying it, but I suppose you tried that and it introduces some other C deps that you have to build?

**Comment by smoser** [comment]
The next trick to drop dependencies would be to just "shell out" to cryptsetup and dmsetup.
Normally I would argue heavily against shelling out.  But I think it is different here for a couple reasons:
 * 'dmsetup' and 'cryptsetup' seem to be reasonable dependencies to me . They are very mature cli interfaces that are very standard on linux.
 * dmsetup and cryptsetup cover the vast majority of the dynamic dependencies that stacker needs to use.

when atomfs builds, it gets a dependency (from ldd) like the following. I marked the ones that come from cryptsetup with a '-'.

```
 /lib64/ld-linux-x86-64.so.2 (0x00007f8a04774000)
-libargon2.so.1 => /lib/x86_64-linux-gnu/libargon2.so.1
-libblkid.so.1 => /lib/x86_64-linux-gnu/libblkid.so.1
-libcrypto.so.3 => /lib/x86_64-linux-gnu/libcrypto.so.3
-libcryptsetup.so.12 => /lib/x86_64-linux-gnu/libcryptsetup.so.12
 libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6
-libdevmapper.so.1.02.1 => /lib/x86_64-linux-gnu/libdevmapper.so.1.02.1
-libjson-c.so.5 => /lib/x86_64-linux-gnu/libjson-c.so.5
 liblz4.so.1 => /lib/x86_64-linux-gnu/liblz4.so.1
 liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 
-libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 
-libpcre2-8.so.0 => /lib/x86_64-linux-gnu/libpcre2-8.so.0 
-libselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 
-libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 
-libuuid.so.1 => /lib/x86_64-linux-gnu/libuuid.so.1 
 libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 
 libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 
 linux-vdso.so.1 
```

So basically if we shelled out, i think we could maybe get atomfs's deps down to only those compression libraries.

(note, stacker probably has others, i was just looking at atomfs now, and it
only imports stackerbuild.io/stacker/pkg/atomfs and
stackerbuild.io/stacker/pkg/mount, but a lot of functionality comes from just that)

**Comment by smoser** [comment]
> My only thought is that it would be nice to import superblock.go instead of copying it, but I suppose you tried that and it introduces some other C deps that you have to build?

All diskfs exposes is a [Filesystem](https://github.com/diskfs/go-diskfs/blob/e29f8c4774c504fb5166defa77214e499e6e9875/filesystem/filesystem.go#L10) interface.  You do a [Read](https://github.com/diskfs/go-diskfs/blob/e29f8c4774c504fb5166defa77214e499e6e9875/filesystem/squashfs/squashfs.go#L100) and get one back, but then you don't have any way to get 'size' or anything even though they're in the struct.

I could propose an upstream change to expose them in a consistent manner for each of the Filesystem interface implementations; this is just lazy.

**Comment by tych0** [comment]
You could downcast to the right one if the individual implentations are public.

**Comment by smoser** [comment]
> You could downcast to the right one if the individual implentations are public.
> […](#)

[squashfs.Filesystem](https://github.com/diskfs/go-diskfs/blob/e29f8c4774c504fb5166defa77214e499e6e9875/filesystem/squashfs/squashfs.go#L23) is what you get back from the Read.   It has a '[superblock]' of type [*superblock](https://github.com/diskfs/go-diskfs/blob/e29f8c4774c504fb5166defa77214e499e6e9875/filesystem/squashfs/superblock.go#L65) member that is not exposed (lower case).  The superblock type has a non-exposed uint64 size.  

Maybe i could manage to get it through an import, but not in any way that was clean or imo preferable to this one file copy.

**Comment by smoser** [comment]
oh for pete's sake. test taht failed is tests/caching.bats "can read previous version's cache".
That test does a clone of stacker and attempts to build it.  I can't build the previous version of stacker because I don't have one of it's dependencies (libsquashfs1 -- thats what i'm tryint to fix).

**Comment by tych0** [comment]
Yeah, you'd have to do reflection, and then you're at risk of some non-public API change. Another option would be to fork + go replace, but copy seems fine too.

**Comment by smoser** [comment]
I think 

> The next trick to drop dependencies would be to just "shell out" to cryptsetup and dmsetup. Normally I would argue heavily against shelling out. But I think it is different here for a couple reasons:
...

I guess there are the following cases
1. kernel 'mount' of a squashfs+dmverity image
2. creation of verity data and root hash

It seems that mount is pretty straightforward with dmsetup or cryptsetup. creation might be a little more difficult.
if we separated use and creation into separate directories, then a consumer (such as 'machine' or 'atomfs') that only needed mount would not end up with the library dep.

File: test/caching.bats [hunk 1]
```diff
@@ -185,6 +185,7 @@ EOF
 }
 
 @test "can read previous version's cache" {
+    skip "old stacker will not build because libsquashfs1-dev is not installed"
```
Comment (rchincha) [inline test/caching.bats:188]: No, we don't want to disable this in general right?
We expect this test alone to fail and we can force-merge.

File: doc/install.md [hunk 2]
```diff
@@ -37,7 +37,7 @@ packages:
 #### **Ubuntu 22.04**
 
     sudo apt install lxc-dev libacl1-dev libgpgme-dev libcap-dev libseccomp-dev
-    sudo apt install libpam0g-dev libselinux-dev libssl-dev libzstd-dev libcryptsetup-dev libdevmapper-dev cryptsetup-bin pkg-config libsquashfs1 libsquashfs-dev
+    sudo apt install libpam0g-dev libselinux-dev libssl-dev libzstd-dev libcryptsetup-dev libdevmapper-dev cryptsetup-bin pkg-config
```
Comment (rchincha) [inline doc/install.md:40]: nice!

File: test/caching.bats (hunk 1 above)
Comment (rchincha) [reply test/caching.bats:188]: The intent I believe is PRs don't break backward compatibility. We can disable this test for now, but we need a new PR to re-enable it right afterwards.

**Review by rchincha — APPROVED** [review]
lgtm
